lib/http/request/writer.rb

```diff
@@ -51,12 +51,16 @@ def connect_through_proxy
       # Adds the headers to the header array for the given request body we are working
       # with
       def add_body_type_headers
-        if @body.is_a?(String) && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
-        elsif @body.nil? && !@headers[Headers::CONTENT_LENGTH]
-          @request_header << "#{Headers::CONTENT_LENGTH}: 0"
-        elsif @body.is_a?(Enumerable) && CHUNKED != @headers[Headers::TRANSFER_ENCODING]
-          raise(RequestError, "invalid transfer encoding")
+        unless @headers[Headers::CONTENT_LENGTH]
+          if @body.is_a?(String)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
+          elsif @body.respond_to?(:read)
+            @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.size}"
```

Comment:
if we are checking `respond_to? :read` we should also check here `respond_to? size`, no? So I believe the best will be:

``` ruby
if @body.respond_to?(:bytesize)
  @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.bytesize}"
elsif @body.respond_to?(:size) && @body.respond_to?(:read)
  @request_header << "#{Headers::CONTENT_LENGTH}: #{@body.size}"
elsif @body.nil?
  @request_header << "#{Headers::CONTENT_LENGTH}: 0"
elsif ...
end
```

@httprb/core please take a look if you see any gotchas here